src/preferences/general.cpp

```diff
@@ -1062,5 +1068,61 @@ void set_achievement(const std::string& content_for, const std::string& id)
 	prefs.add_child("achievements", ach);
 }
 
+int progress_achievement(const std::string& content_for, const std::string& id, int limit, int max_progress, int amount)
+{
+	bool completed = achievement(content_for, id);
+
+	if(!completed)
+	{
+		for(config& ach : prefs.child_range("achievements"))
+		{
+			// if achievements already exist for this content and the achievement has not already been set, add it
+			if(ach["content_for"].str() == content_for)
+			{
+				// check if this achievement has progressed before - if so then increment it
+				for(config& in_progress : ach.child_range("in_progress"))
+				{
+					if(in_progress["id"].str() == id)
+					{
+						in_progress["progress_at"] = std::clamp(in_progress["progress_at"].to_int() + amount, 0, std::min(limit, max_progress));
+						return in_progress["progress_at"].to_int();
+					}
+				}
+
+				// else this is the first time this achievement is progressing
+				if(amount != 0)
+				{
+					config set_progress;
+					set_progress["id"] = id;
+					set_progress["progress_at"] = std::clamp(amount, 0, std::min(limit, max_progress));
+
+					config& child = ach.add_child("in_progress", set_progress);
+					return child["progress_at"].to_int();
+				}
+				return 0;
+			}
+		}
+
+		// else not only has this achievement not progressed before, this is the first achievement for this achievement group to be added
+		if(amount != 0)
+		{
+			config ach;
+			config set_progress;
+
+			set_progress["id"] = id;
+			set_progress["progress_at"] = std::clamp(amount, 0, std::min(limit, max_progress));
+
+			ach["content_for"] = content_for;
+			ach["ids"] = "";
+
+			config& child = ach.add_child("in_progress", set_progress);
+			prefs.add_child("achievements", ach);
+			return child["progress_at"].to_int();
+		}
+		return 0;
+	}
+
+	return -1;
```

Comment:
Most of this function could have one less indent, by changing the block around line 1075 to be
```
if](completed) {
    return -1;
}
````

This function has other early returns, so having an early return right at the top seems reasonable.

Reply:
Updated.